ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/DebugDataDumper.java

```diff
@@ -0,0 +1,143 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class DebugDataDumper {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final String GOSSIP_MESSAGES_DIR = "gossip_messages";
+  private static final String DECODING_ERROR_SUB_DIR = "decoding_error";
+  private static final String REJECTED_SUB_DIR = "rejected";
+  private static final String INVALID_BLOCK_DIR = "invalid_blocks";
+
+  private boolean enabled;
+  private final Path directory;
+
+  public DebugDataDumper(final Path directory, final boolean enabled) {
+    this.enabled = enabled;
+    this.directory = directory;
+    if (!enabled) {
+      return;
+    }
+
+    final Path gossipMessagesPath = this.directory.resolve(GOSSIP_MESSAGES_DIR);
+    createDirectory(gossipMessagesPath, GOSSIP_MESSAGES_DIR, "gossip messages");
+    createDirectory(
+        gossipMessagesPath.resolve(DECODING_ERROR_SUB_DIR),
+        DECODING_ERROR_SUB_DIR,
+        "gossip messages with decoding errors");
+    createDirectory(
+        gossipMessagesPath.resolve(REJECTED_SUB_DIR), REJECTED_SUB_DIR, "rejected gossip messages");
+    createDirectory(this.directory.resolve(INVALID_BLOCK_DIR), INVALID_BLOCK_DIR, "invalid blocks");
+  }
+
+  public void saveGossipMessageDecodingError(
+      final String topic, final String arrivalTimestamp, final Bytes originalMessage) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final Path topicPath =
+        Path.of(GOSSIP_MESSAGES_DIR).resolve(DECODING_ERROR_SUB_DIR).resolve(topic);
+    saveBytesToFile(
+        "gossip message with decoding error", topicPath.resolve(fileName), originalMessage);
+  }
+
+  public void saveGossipRejectedMessageToFile(
+      final String topic, final String arrivalTimestamp, final Bytes decodedMessage) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final Path topicPath = Path.of(GOSSIP_MESSAGES_DIR).resolve(REJECTED_SUB_DIR).resolve(topic);
+    saveBytesToFile("rejected gossip message", topicPath.resolve(fileName), decodedMessage);
+  }
+
+  public void saveInvalidBlockToFile(
+      final UInt64 slot, final Bytes32 blockRoot, final Bytes blockSsz) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName =
+        String.format("slot%s_root%s.ssz", slot, blockRoot.toUnprefixedHexString());
```

Comment:
i would probably drop the prefix wordings, they're going to be very long filenames as it is, and this is just noise most likely, unless we have a really good reason for them